c_tests/tests/noopts.c

```diff
@@ -7,67 +7,52 @@
 //     ...
 //     store i32 2...
 //     ...
-//     store i32 3...
-//     ...
 //     ...add...
 //     ...
 //     --- End aot ---
 //     --- Begin jit-pre-opt ---
 //     ...
 //     store i32 2...
 //     ...
-//     store i32 3...
-//     ...
 //     ...add...
 //     ...
 //     --- End jit-pre-opt ---
 //     --- Begin jit-post-opt ---
 //     ...
 //     store i32 2...
 //     ...
-//     store i32 3...
-//     ...
 //     ...add...
 //     ...
 //     --- End jit-post-opt ---
 //     ...
 
-// Check that our NOOPT macro blocks optimisations.
+// Check that our NOOPT_VAL macro blocks optimisations.
 
 #include <assert.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include <yk_testing.h>
 
-int add_them(int x, int y) {
-  // At the call-site x and y are constant, so an optimising compiler would try
-  // to evaluate `x + y` at compile time, replacing the result with a constant
-  // value. Our NOOPT macro should block this.
-  return x + y;
-}
-
 int main(int argc, char **argv) {
-  int x, y, res;
-  void *tt = __yktrace_start_tracing(HW_TRACING, &res, &x, &y);
-  // The __yktrace_start_tracing() call has almost certainly already blocked
-  // optimisation of x and y, as pointers to them are passed as arguments and
-  // the compiler must assume that it may mutate them. Therefore to properly
-  // test NOOPT we need initialise x and y after the call.
+  int x, res;
+  void *tt = __yktrace_start_tracing(HW_TRACING, &res, &x);
+  // __yktrace_start_tracing() will already inhibit optimisations on x since
+  // it's address is passed as an argument and the compiler must assume that it
+  // may be mutated. Therefore to properly test NOOPT_VAL we need initialise x
+  // and y after the call to __yktrace_start_tracing..
   x = 2;
-  NOOPT(&x);
-  y = 3;
-  NOOPT(&y);
-  res = add_them(x, y);
+  NOOPT_VAL(x);
+  res = x + 3; // We don't want this operation to be optimised away.
```

Comment:
I claim that `NOOPT_VAL(res)` is not necessary in this example because in `res = x + 3`, `x` must be opaque.

Reply:
It's just convention to guard the inputs and the outputs of the computation you want to shield from optimisations:

> The key to these techniques is to focus on the data. You make the input to the computation opaque to the optimizer and the result of the computation opaque to the optimizer.

https://stackoverflow.com/questions/11748362/how-to-disable-an-optimization-in-llvm

I don't mind either way, it's not strictly required for this particular test on Linux today.

Reply:
I'm not sure that SO link is the right one?

Reply:
Oops! https://stackoverflow.com/questions/37786547/enforcing-statement-order-in-c/38025837#38025837

Reply:
You're right and I'm wrong: it's not that we need the second `NOOPT_VAL` *as such* but putting it before `stop_tracing` ensures that `res` won't float past `stop_tracing`.

[FWIW, `res` seems a less than ideal variable name in many of these tests. It would probably be better just to use obviously meaningless names like `y`. Maybe we can think about that in a future PR?]